Rendering update

JeanCGedeon opened the pull request:
fix fetch request on the front-end
add userId into state on the front-end
edit the top bar
add attractionReviewsRouter(also update userId in the post request)

Commit:
work in progress, debugging

camziny committed:
created "attractionReviews" router

camziny committed:
imported attractionReviews into attractionRouter

Commit:
work in progress, placeholder locationId, rendering attraction form

Commit:
Add attraction reviews router

Co-authored-by: JeanCGedeon <JeanCGedeon@users.noreply.github.com>
Co-authored-by: Tyler <br.ndt.tyler@gmail.com>
Co-authored-by: camziny <camziny@users.noreply.github.com>

Commit:
Beginning of work on rendering new reviews

JeanCGedeon committed:
modify top bar
fix front-end post request for reviews (added userId in state)
merge attractionReviewsRouter
merge from
https://github.com/br-ndt/trip/commit/4232cef0de79a4bf2d4408fc68b2e9a9fa9ea903

Co-authored-by: Tyler <br.ndt.tyler@gmail.com>
Co-authored-by: camziny <camziny@users.noreply.github.com>

nwalberts commented on client/src/components/App.js:
```diff
@@ -31,9 +31,10 @@ const App = (props) => {
       <TopBar user={currentUser} />
+      <Route path="/">
+        <h1>Trip</h1>
+      </Route>
```
This would be considered an incomplete feature. Ill leave for now because your app came with something similar, but we really should just delete this until we have something for this page

br-ndt replied:
Changed the rendered component to be a Redirect to "/attractions"

nwalberts commented on client/src/components/AttractionShowPage.js:
```diff
@@ -38,11 +41,18 @@ const AttractionShowPage = (props) => {
   });
 
+  const attractionName = attraction.name || "Launch Academy";
+  const attractionDescription = attraction.description || "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excep ...
```
both of these are considered "filler text" and mark an incomplete feature. These elements either shouldnt appear on the page at all, or be based on data from your backend. No filler

br-ndt replied:
Changed these filler/hardcode values to be conditionally-rendered elements.

nwalberts commented on client/src/components/AttractionShowPage.js:
```diff
@@ -38,11 +41,18 @@ const AttractionShowPage = (props) => {
+      <div>
+        <h4>Add a Review</h4>
+        <NewReviewForm attractionId={id} addNewReview={addNewReview}/>
+      </div>
```
The div and h4 here are part of what makes up the form. Though the div currently so serves no purpose
- delete the div entirely 
- move the h4 into the NewReviewForm

nwalberts commented on client/src/components/AttractionShowPage.js:
```diff
@@ -30,6 +29,10 @@ const AttractionShowPage = (props) => {
 
+  const addNewReview = (review) => {
+    setAttraction({...attraction, reviews: [...attraction.reviews, review]});
+  }
```
yooo this is an incomplete feature. If we add this review, it wont get persisted in the database. This should post to the backend and persist the features. No half measures please!

nwalberts replied:
I misspoke. The names you have here are confusing. You have one addNewReview method in the form, and one here. Let's disambiguate these two and rename either this one or the one with the POST

br-ndt replied:
Function names changed to postAttraction/postReview

nwalberts commented on client/src/components/AttractionsList.js:
```diff
@@ -29,7 +34,12 @@ const AttractionsList = (props) => {
+      <h3>Add a New Attraction:</h3>
+      <div>
+        <NewAttractionForm addNewAttraction={addNewAttraction}/>
+      </div>
```
this div is serving no purpose, please delete

nwalberts replied:
Im pretty confused, because in your form, you redirect to the AttractionList page....but you are already on the attraction list page

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -8,6 +8,8 @@ const NewAttractionForm = (props) => {
     description: "",
+    // placeholder while we work on locationId integration
+    locationId: 2,
   });
```
placeholders are not allowed, this marks another incomplete feature. Either have this independent of a location or find another way

nwalberts replied:
Furthermore, you shouldnt be storing a locationId in state in the first place. It should be something passed down via props

nwalberts replied:
This newAttractionForm either needs to be on the LocationShowPage, or on its own page. Otherwise, we have no idea which location this is being added to. 

alterantively, you can have an attractions have an _optional_ BelongsToOne relationship with location, but that would require some updates

nwalberts replied:
As we talked, we also said you could have a dropdown of locations to set for this new attraction. In which case, the locationId CAN be tracked in state

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -27,13 +29,16 @@ const NewAttractionForm = (props) => {
       } else {
-        setShouldRedirect(true);
+        //clearForm();
+        props.addNewAttraction(body.attraction);
```
no code comments, remove line 32

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -27,13 +29,16 @@ const NewAttractionForm = (props) => {
   };
 
+  
+
```
....we dont need these extra lines. Please delete

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -50,7 +55,6 @@ const NewAttractionForm = (props) => {
 
```
yo, you all need to decide, if this form lives on the index page and if the form submission adds the new attraction to state (and posts) or if this form lives on its own page and redirects back to attractions. Im rather confused

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -50,7 +55,6 @@ const NewAttractionForm = (props) => {
 
   return (
     <div>
-      <h3>Add a New Attraction</h3>
```
I actually liked that h3 here...

Also the div on line 57 is serving no purpose, please delete

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -50,7 +55,6 @@ const NewAttractionForm = (props) => {
-      <h3>Add a New Attraction</h3>
       <ErrorList errors={errors}/>
       <form className="attraction-form form" onSubmit={handleSubmit}>
         <input type="text" name="name" placeholder="The Colosseum" onChange={handleInputChange}/>
```
all of the placeholders in this form should be deleted, they don't inform the user on how to fill out the form, and would confuse users and emplpyers

nwalberts replied:
maybe something a little more relatable ::P

nwalberts commented on client/src/components/NewReviewForm.js:
```diff
@@ -1,12 +1,37 @@
     rating: "",
     content: "",
+    userId: 1,// Placeholder value
   });
```
No placeholder values, and userId should not be tracked in state here. It should be passed down via props from App.js

nwalberts commented on client/src/components/NewReviewForm.js:
```diff
@@ -17,28 +42,27 @@ const NewReviewForm = ({ postReview }) => {
       title: "",
       rating: "",
       content: "",
+      userId: 1
```
delete

nwalberts commented on package.json:
```diff
@@ -18,6 +18,6 @@
   "engines": {
     "node": "^14.19",
-    "yarn": "^1.22.18"
+    "yarn": "^1.22.17"
```
is there a reason for the yarn change? Did you discuss this as a group?

nwalberts commented on server/src/routes/api/v1/attractionReviewsRouter.js:
```diff
@@ -0,0 +1,19 @@
+  } catch (error) {
+    return res.status(500).json({ errors: error });
+  }
+});
```
user id should come from your passport session, via `req.user`, not from the frontend.

nwalberts replied:
there is also no ValidationError handling here. You are handling for these types of errors on the frontend, but in this implementation those errors would never be used

nwalberts commented on client/src/components/NewReviewForm.js:
```diff
@@ -1,12 +1,37 @@
+      if (!response.ok) {
+        if (response.status === 422) {
+          const newErrors = translateServerErrors(body.errors);
+          return setErrors(newErrors);
```
wait...where is `setErrors` defined? Are those errors ever used and displayed anywhere?

nwalberts commented on client/src/components/NewAttractionForm.js:
```diff
@@ -27,13 +29,16 @@ const NewAttractionForm = (props) => {
         }
```
you are going to need to update this post fetch request. Now that you have a location, the path should be nested 
```
fetch("/api/v1/locations/:locationId/attractions", ...)
```

nwalberts replied:
alternatively, you can
- add a dropdown with locations 
- add this entire form to the LocationShow
- make this its in own page

br-ndt replied:
Added location dropdown list to NewAttractionForm
Added error handling for this new POST route, located in locationAttractionsRoute

nwalberts commented on server/src/routes/api/v1/attractionsRouter.js:
```diff
@@ -15,11 +19,12 @@ attractionsRouter.get("/", async (req, res) => {
 
 attractionsRouter.post("/", async (req, res) => {
-  const { name, description } = cleanUserInput(req.body);
+  const { name, description, locationId } = cleanUserInput(req.body);
```
the locationId shouldnt be in the body. Per my comment in your form, you should have locationId in your params, not the body. 

You are going to need to move this post into an locationAttractions router

nwalberts commented on server/src/routes/api/v1/attractionsRouter.js:
```diff
@@ -15,11 +19,12 @@ attractionsRouter.get("/", async (req, res) => {
+    const newAttraction = await Attraction.query().insertAndFetch({ name, description, locationId });
     return res.status(201).json({ attraction: newAttraction });
   } catch (error) {
+    console.log(error);
```
remove this

nwalberts commented on client/src/components/AttractionsList.js:
```diff
@@ -19,6 +20,10 @@ const AttractionsList = (props) => {
 
+  const addNewAttraction = (attraction) => {
+    setAttractions([...attractions, attraction]);
+  }
```
this is used in another `AddNewAttraction` method which is very confusing, lets rename one of them (probably the method that has the POST

br-ndt committed:
address @nwalberts feedback:
add locationAttractionsRouter
remove placeholders from forms
incorporate user session data
add authenticated routes for forms
add Location dropdown to NewAttractionForm

Co-authored-by: JeanCGedeon <JeanCGedeon@users.noreply.github.com>
